File: packages/zynth-android/android/ZynthKit/src/main/java/com/zynth/kit/core/ZynthColorParser.kt
package com.zynth.kit.core

import android.graphics.Color
import androidx.core.graphics.ColorUtils
import java.util.Locale
import kotlin.math.roundToInt

object ZynthColorParser {
    private val NAMED_COLORS = mapOf(
        "aliceblue" to 0xfff0f8ff.toInt(),
        "antiquewhite" to 0xfffaebd7.toInt(),
        "aqua" to 0xff00ffff.toInt(),
        "aquamarine" to 0xff7fffd4.toInt(),
        "azure" to 0xfff0ffff.toInt(),
        "beige" to 0xfff5f5dc.toInt(),
        "bisque" to 0xffffe4c4.toInt(),
        "black" to 0xff000000.toInt(),
        "blanchedalmond" to 0xffffebcd.toInt(),
        "blue" to 0xff0000ff.toInt(),
        "blueviolet" to 0xff8a2be2.toInt(),
        "brown" to 0xffa52a2a.toInt(),
        "burlywood" to 0xffdeb887.toInt(),
        "cadetblue" to 0xff5f9ea0.toInt(),
        "chartreuse" to 0xff7fff00.toInt(),
        "chocolate" to 0xffd2691e.toInt(),
        "coral" to 0xffff7f50.toInt(),
        "cornflowerblue" to 0xff6495ed.toInt(),
        "cornsilk" to 0xfffff8dc.toInt(),
        "crimson" to 0xffdc143c.toInt(),
        "cyan" to 0xff00ffff.toInt(),
        "darkblue" to 0xff00008b.toInt(),
        "darkcyan" to 0xff008b8b.toInt(),
        "darkgoldenrod" to 0xffb8860b.toInt(),
        "darkgray" to 0xffa9a9a9.toInt(),
        "darkgreen" to 0xff006400.toInt(),
        "darkgrey" to 0xffa9a9a9.toInt(),
        "darkkhaki" to 0xffbdb76b.toInt(),
        "darkmagenta" to 0xff8b008b.toInt(),
        "darkolivegreen" to 0xff556b2f.toInt(),
        "darkorange" to 0xffff8c00.toInt(),
        "darkorchid" to 0xff9932cc.toInt(),
        "darkred" to 0xff8b0000.toInt(),
        "darksalmon" to 0xffe9967a.toInt(),
        "darkseagreen" to 0xff8fbc8f.toInt(),
        "darkslateblue" to 0xff483d8b.toInt(),
        "darkslategrey" to 0xff2f4f4f.toInt(),
        "darkturquoise" to 0xff00ced1.toInt(),
        "darkviolet" to 0xff9400d3.toInt(),
        "deeppink" to 0xffff1493.toInt(),
        "deepskyblue" to 0xff00bfff.toInt(),
        "dimgray" to 0xff696969.toInt(),
        "dimgrey" to 0xff696969.toInt(),
        "dodgerblue" to 0xff1e90ff.toInt(),
        "firebrick" to 0xffb22222.toInt(),
        "floralwhite" to 0xfffffaf0.toInt(),
        "forestgreen" to 0xff228b22.toInt(),
        "fuchsia" to 0xffff00ff.toInt(),
        "gainsboro" to 0xffdcdcdc.toInt(),
        "ghostwhite" to 0xfff8f8ff.toInt(),
        "gold" to 0xffffd700.toInt(),
        "goldenrod" to 0xffdaa520.toInt(),
        "gray" to 0xff808080.toInt(),
        "green" to 0xff008000.toInt(),
        "greenyellow" to 0xffadff2f.toInt(),
        "grey" to 0xff808080.toInt(),
        "honeydew" to 0xfff0fff0.toInt(),
        "hotpink" to 0xffff69b4.toInt(),
        "indianred" to 0xffcd5c5c.toInt(),
        "indigo" to 0xff4b0082.toInt(),
        "ivory" to 0xfffffff0.toInt(),
        "khaki" to 0xfff0e68c.toInt(),
        "lavender" to 0xffe6e6fa.toInt(),
        "lavenderblush" to 0xfffff0f5.toInt(),
        "lawngreen" to 0xff7cfc00.toInt(),
        "lemonchiffon" to 0xfffffacd.toInt(),
        "lightblue" to 0xffadd8e6.toInt(),
        "lightcoral" to 0xfff08080.toInt(),
        "lightcyan" to 0xffe0ffff.toInt(),
        "lightgoldenrodyellow" to 0xfffafad2.toInt(),
        "lightgray" to 0xffd3d3d3.toInt(),
        "lightgreen" to 0xff90ee90.toInt(),
        "lightgrey" to 0xffd3d3d3.toInt(),
        "lightpink" to 0xffffb6c1.toInt(),
        "lightsalmon" to 0xffffa07a.toInt(),
        "lightseagreen" to 0xff20b2aa.toInt(),
        "lightskyblue" to 0xff87cefa.toInt(),
        "lightslategrey" to 0xff778899.toInt(),
        "lightsteelblue" to 0xffb0c4de.toInt(),
        "lightyellow" to 0xffffffe0.toInt(),
        "lime" to 0xff00ff00.toInt(),
        "limegreen" to 0xff32cd32.toInt(),
        "linen" to 0xfffaf0e6.toInt(),
        "magenta" to 0xffff00ff.toInt(),
        "maroon" to 0xff800000.toInt(),
        "mediumaquamarine" to 0xff66cdaa.toInt(),
        "mediumblue" to 0xff0000cd.toInt(),
        "mediumorchid" to 0xffba55d3.toInt(),
        "mediumpurple" to 0xff9370db.toInt(),
        "mediumseagreen" to 0xff3cb371.toInt(),
        "mediumslateblue" to 0xff7b68ee.toInt(),
        "mediumspringgreen" to 0xff00fa9a.toInt(),
        "mediumturquoise" to 0xff48d1cc.toInt(),
        "mediumvioletred" to 0xffc71585.toInt(),
        "midnightblue" to 0xff191970.toInt(),
        "mintcream" to 0xfff5fffa.toInt(),
        "mistyrose" to 0xffffe4e1.toInt(),
        "moccasin" to 0xffffe4b5.toInt(),
        "navajowhite" to 0xffffdead.toInt(),
        "navy" to 0xff000080.toInt(),
        "oldlace" to 0xfffdf5e6.toInt(),
        "olive" to 0xff808000.toInt(),
        "olivedrab" to 0xff6b8e23.toInt(),
        "orange" to 0xffffa500.toInt(),
        "orangered" to 0xffff4500.toInt(),
        "orchid" to 0xffda70d6.toInt(),
        "palegoldenrod" to 0xffeee8aa.toInt(),
        "palegreen" to 0xff98fb98.toInt(),
        "paleturquoise" to 0xffafeeee.toInt(),
        "palevioletred" to 0xffdb7093.toInt(),
        "papayawhip" to 0xffffefd5.toInt(),
        "peachpuff" to 0xffffdab9.toInt(),
        "peru" to 0xffcd853f.toInt(),
        "pink" to 0xffffc0cb.toInt(),
        "plum" to 0xffdda0dd.toInt(),
        "powderblue" to 0xffb0e0e6.toInt(),
        "purple" to 0xff800080.toInt(),
        "rebeccapurple" to 0xff663399.toInt(),
        "red" to 0xffff0000.toInt(),
        "rosybrown" to 0xffbc8f8f.toInt(),
        "royalblue" to 0xff4169e1.toInt(),
        "saddlebrown" to 0xff8b4513.toInt(),
        "salmon" to 0xfffa8072.toInt(),
        "sandybrown" to 0xfff4a460.toInt(),
        "seagreen" to 0xff2e8b57.toInt(),
        "seashell" to 0xfffff5ee.toInt(),
        "sienna" to 0xffa0522d.toInt(),
        "silver" to 0xffc0c0c0.toInt(),
        "skyblue" to 0xff87ceeb.toInt(),
        "slateblue" to 0xff6a5acd.toInt(),
        "slategray" to 0xff708090.toInt(),
        "snow" to 0xfffffafa.toInt(),
        "springgreen" to 0xff00ff7f.toInt(),
        "steelblue" to 0xff4682b4.toInt(),
        "tan" to 0xffd2b48c.toInt(),
        "teal" to 0xff008080.toInt(),
        "thistle" to 0xffd8bfd8.toInt(),
        "tomato" to 0xffff6347.toInt(),
        "turquoise" to 0xff40e0d0.toInt(),
        "violet" to 0xffee82ee.toInt(),
        "wheat" to 0xfff5deb3.toInt(),
        "white" to 0xffffffff.toInt(),
        "whitesmoke" to 0xfff5f5f5.toInt(),
        "yellow" to 0xffffff00.toInt(),
        "yellowgreen" to 0xff9acd32.toInt(),
        "transparent" to 0x00000000
    )

    // rgb(r, g, b) or rgba(r, g, b, a)
    // Supports spaces, commas, and optional alpha
    private val RGB_REGEX = Regex("""^rgba?\(\s*([0-9]+)\s*(?:,|\s)\s*([0-9]+)\s*(?:,|\s)\s*([0-9]+)\s*(?:[,/]\s*([0-9]*\.?[0-9]+%?))?\s*\)$""", RegexOption.IGNORE_CASE)
    
    // hsl(h, s, l) or hsla(h, s, l, a)
    private val HSL_REGEX = Regex("""^hsla?\(\s*([0-9]+(?:deg)?)\s*(?:,|\s)\s*([0-9.]+%)\s*(?:,|\s)\s*([0-9.]+%)\s*(?:[,/]\s*([0-9]*\.?[0-9]+%?))?\s*\)$""", RegexOption.IGNORE_CASE)
    
    // hwb(h, w, b, a?)
    private val HWB_REGEX = Regex("""^hwb\(\s*([0-9]+(?:deg)?)\s*(?:,|\s)\s*([0-9.]+%)\s*(?:,|\s)\s*([0-9.]+%)\s*(?:[,/]\s*([0-9]*\.?[0-9]+%?))?\s*\)$""", RegexOption.IGNORE_CASE)

    fun parse(color: String?): Int? {
        if (color.isNullOrBlank()) return null
        val trimmed = color.trim()

        // 1. Hex Color
        if (trimmed.startsWith("#")) {
            return parseHex(trimmed)
        }

        // 2. Named Color
        val lower = trimmed.lowercase(Locale.ROOT)
        if (NAMED_COLORS.containsKey(lower)) {
            return NAMED_COLORS[lower]
        }

        // 3. Functional Syntax
        return when {
            trimmed.startsWith("rgb", ignoreCase = true) -> parseRgb(trimmed)
            trimmed.startsWith("hsl", ignoreCase = true) -> parseHsl(trimmed)
            trimmed.startsWith("hwb", ignoreCase = true) -> parseHwb(trimmed)
            else -> null
        }
    }

    private fun parseHex(hex: String): Int? {
        var raw = hex.substring(1)
        return try {
            when (raw.length) {
                3 -> {
                    // #RGB -> #RRGGBB
                    val r = raw[0]
                    val g = raw[1]
                    val b = raw[2]
                    Color.parseColor("#$r$r$g$g$b$b")
                }
                4 -> {
                    // #RGBA -> #RRGGBBAA
                    val r = raw[0]
                    val g = raw[1]
                    val b = raw[2]
                    val a = raw[3]
                    Color.parseColor("#$a$a$r$r$g$g$b$b")
                }
                6 -> {
                    // #RRGGBB
                    Color.parseColor(hex)
                }
                8 -> {
                    // #RRGGBBAA in Web/CSS
                    // Android Color.parseColor expects #AARRGGBB
                    // We need to rotate: RRGGBBAA -> AARRGGBB
                    val r = raw.substring(0, 2)
                    val g = raw.substring(2, 4)
                    val b = raw.substring(4, 6)
                    val a = raw.substring(6, 8)
                    Color.parseColor("#$a$r$g$b")
                }
                else -> null
            }
        } catch (e: IllegalArgumentException) {
            null
        }
    }

    private fun parseRgb(value: String): Int? {
        val match = RGB_REGEX.matchEntire(value) ?: return null
        val (rStr, gStr, bStr) = match.destructured
        val aStr = match.groups[4]?.value

        val r = rStr.toInt().coerceIn(0, 255)
        val g = gStr.toInt().coerceIn(0, 255)
        val b = bStr.toInt().coerceIn(0, 255)
        val a = parseAlpha(aStr)

        return Color.argb(a, r, g, b)
    }

    private fun parseHsl(value: String): Int? {
        val match = HSL_REGEX.matchEntire(value) ?: return null
        val (hStr, sStr, lStr) = match.destructured
        val aStr = match.groups[4]?.value

        val h = parseDegrees(hStr)
        val s = parsePercentage(sStr)
        val l = parsePercentage(lStr)
        val a = parseAlpha(aStr)

        return ColorUtils.HSLToColor(floatArrayOf(h, s, l)).let {
             Color.argb(a, Color.red(it), Color.green(it), Color.blue(it))
        }
    }
    
    private fun parseHwb(value: String): Int? {
        val match = HWB_REGEX.matchEntire(value) ?: return null
        val (hStr, wStr, bStr) = match.destructured
        val aStr = match.groups[4]?.value
        
        val h = parseDegrees(hStr)
        val w = parsePercentage(wStr)
        val b = parsePercentage(bStr)
        val a = parseAlpha(aStr)
        
        if (w + b >= 1f) {
            val gray = ((w / (w + b)) * 255).roundToInt()
            return Color.argb(a, gray, gray, gray)
        }
        
        // HWB to RGB conversion
        // 1. Get RGB from HSL with (H, 100%, 50%) -> Pure saturated color
        val rgb = ColorUtils.HSLToColor(floatArrayOf(h, 1f, 0.5f))
        val rBase = Color.red(rgb) / 255f
        val gBase = Color.green(rgb) / 255f
        val bBase = Color.blue(rgb) / 255f
        
        // 2. Scale by whiteness and blackness
        // rgb = rgb * (1 - w - b) + w
        val factor = 1f - w - b
        val r = ((rBase * factor) + w) * 255
        val g = ((gBase * factor) + w) * 255
        val bl = ((bBase * factor) + w) * 255
        
        return Color.argb(a, r.roundToInt(), g.roundToInt(), bl.roundToInt())
    }

    private fun parseAlpha(alphaStr: String?): Int {
        if (alphaStr.isNullOrBlank()) return 255
        return if (alphaStr.endsWith("%")) {
            (alphaStr.dropLast(1).toFloat() / 100f * 255).roundToInt().coerceIn(0, 255)
        } else {
            (alphaStr.toFloat() * 255).roundToInt().coerceIn(0, 255)
        }
    }
    
    private fun parsePercentage(str: String): Float {
        return if (str.endsWith("%")) {
             str.dropLast(1).toFloat() / 100f
        } else {
             0f // HSL/HWB usually require % for S/L/W/B in CSS, but we can be lenient or strict
        }.coerceIn(0f, 1f)
    }

    private fun parseDegrees(str: String): Float {
        val raw = str.lowercase().removeSuffix("deg")
        return raw.toFloat() % 360f
    }
}

File: packages/zynth-core/ios/ZynthKit/src/runtime/ZynthColorParser.m
#import "ZynthColorParser.h"

@implementation ZynthColorParser

static NSDictionary<NSString *, NSNumber *> *ZynthNamedColors;

+ (void)initialize {
  if (self == [ZynthColorParser class]) {
    ZynthNamedColors = @{
      @"aliceblue": @0xf0f8ff,
      @"antiquewhite": @0xfaebd7,
      @"aqua": @0x00ffff,
      @"aquamarine": @0x7fffd4,
      @"azure": @0xf0ffff,
      @"beige": @0xf5f5dc,
      @"bisque": @0xffe4c4,
      @"black": @0x000000,
      @"blanchedalmond": @0xffebcd,
      @"blue": @0x0000ff,
      @"blueviolet": @0x8a2be2,
      @"brown": @0xa52a2a,
      @"burlywood": @0xdeb887,
      @"cadetblue": @0x5f9ea0,
      @"chartreuse": @0x7fff00,
      @"chocolate": @0xd2691e,
      @"coral": @0xff7f50,
      @"cornflowerblue": @0x6495ed,
      @"cornsilk": @0xfff8dc,
      @"crimson": @0xdc143c,
      @"cyan": @0x00ffff,
      @"darkblue": @0x00008b,
      @"darkcyan": @0x008b8b,
      @"darkgoldenrod": @0xb8860b,
      @"darkgray": @0xa9a9a9,
      @"darkgreen": @0x006400,
      @"darkgrey": @0xa9a9a9,
      @"darkkhaki": @0xbdb76b,
      @"darkmagenta": @0x8b008b,
      @"darkolivegreen": @0x556b2f,
      @"darkorange": @0xff8c00,
      @"darkorchid": @0x9932cc,
      @"darkred": @0x8b0000,
      @"darksalmon": @0xe9967a,
      @"darkseagreen": @0x8fbc8f,
      @"darkslateblue": @0x483d8b,
      @"darkslategrey": @0x2f4f4f,
      @"darkturquoise": @0x00ced1,
      @"darkviolet": @0x9400d3,
      @"deeppink": @0xff1493,
      @"deepskyblue": @0x00bfff,
      @"dimgray": @0x696969,
      @"dimgrey": @0x696969,
      @"dodgerblue": @0x1e90ff,
      @"firebrick": @0xb22222,
      @"floralwhite": @0xfffaf0,
      @"forestgreen": @0x228b22,
      @"fuchsia": @0xff00ff,
      @"gainsboro": @0xdcdcdc,
      @"ghostwhite": @0xf8f8ff,
      @"gold": @0xffd700,
      @"goldenrod": @0xdaa520,
      @"gray": @0x808080,
      @"green": @0x008000,
      @"greenyellow": @0xadff2f,
      @"grey": @0x808080,
      @"honeydew": @0xf0fff0,
      @"hotpink": @0xff69b4,
      @"indianred": @0xcd5c5c,
      @"indigo": @0x4b0082,
      @"ivory": @0xfffff0,
      @"khaki": @0xf0e68c,
      @"lavender": @0xe6e6fa,
      @"lavenderblush": @0xfff0f5,
      @"lawngreen": @0x7cfc00,
      @"lemonchiffon": @0xfffacd,
      @"lightblue": @0xadd8e6,
      @"lightcoral": @0xf08080,
      @"lightcyan": @0xe0ffff,
      @"lightgoldenrodyellow": @0xfafad2,
      @"lightgray": @0xd3d3d3,
      @"lightgreen": @0x90ee90,
      @"lightgrey": @0xd3d3d3,
      @"lightpink": @0xffb6c1,
      @"lightsalmon": @0xffa07a,
      @"lightseagreen": @0x20b2aa,
      @"lightskyblue": @0x87cefa,
      @"lightslategrey": @0x778899,
      @"lightsteelblue": @0xb0c4de,
      @"lightyellow": @0xffffe0,
      @"lime": @0x00ff00,
      @"limegreen": @0x32cd32,
      @"linen": @0xfaf0e6,
      @"magenta": @0xff00ff,
      @"maroon": @0x800000,
      @"mediumaquamarine": @0x66cdaa,
      @"mediumblue": @0x0000cd,
      @"mediumorchid": @0xba55d3,
      @"mediumpurple": @0x9370db,
      @"mediumseagreen": @0x3cb371,
      @"mediumslateblue": @0x7b68ee,
      @"mediumspringgreen": @0x00fa9a,
      @"mediumturquoise": @0x48d1cc,
      @"mediumvioletred": @0xc71585,
      @"midnightblue": @0x191970,
      @"mintcream": @0xf5fffa,
      @"mistyrose": @0xffe4e1,
      @"moccasin": @0xffe4b5,
      @"navajowhite": @0xffdead,
      @"navy": @0x000080,
      @"oldlace": @0xfdf5e6,
      @"olive": @0x808000,
      @"olivedrab": @0x6b8e23,
      @"orange": @0xffa500,
      @"orangered": @0xff4500,
      @"orchid": @0xda70d6,
      @"palegoldenrod": @0xeee8aa,
      @"palegreen": @0x98fb98,
      @"paleturquoise": @0xafeeee,
      @"palevioletred": @0xdb7093,
      @"papayawhip": @0xffefd5,
      @"peachpuff": @0xffdab9,
      @"peru": @0xcd853f,
      @"pink": @0xffc0cb,
      @"plum": @0xdda0dd,
      @"powderblue": @0xb0e0e6,
      @"purple": @0x800080,
      @"rebeccapurple": @0x663399,
      @"red": @0xff0000,
      @"rosybrown": @0xbc8f8f,
      @"royalblue": @0x4169e1,
      @"saddlebrown": @0x8b4513,
      @"salmon": @0xfa8072,
      @"sandybrown": @0xf4a460,
      @"seagreen": @0x2e8b57,
      @"seashell": @0xfff5ee,
      @"sienna": @0xa0522d,
      @"silver": @0xc0c0c0,
      @"skyblue": @0x87ceeb,
      @"slateblue": @0x6a5acd,
      @"slategray": @0x708090,
      @"snow": @0xfffafa,
      @"springgreen": @0x00ff7f,
      @"steelblue": @0x4682b4,
      @"tan": @0xd2b48c,
      @"teal": @0x008080,
      @"thistle": @0xd8bfd8,
      @"tomato": @0xff6347,
      @"turquoise": @0x40e0d0,
      @"violet": @0xee82ee,
      @"wheat": @0xf5deb3,
      @"white": @0xffffffff,
      @"whitesmoke": @0xf5f5f5,
      @"yellow": @0xffff00,
      @"yellowgreen": @0x9acd32,
      @"transparent": @-1
    };
  }
}

+ (nullable UIColor *)parseColor:(nullable NSString *)colorString {
  if (!colorString || colorString.length == 0) return nil;
  NSString *trimmed = [colorString stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
  if (trimmed.length == 0) return nil;

  if ([trimmed hasPrefix:@"#"]) {
    return [self parseHex:trimmed];
  }

  NSString *lower = [trimmed lowercaseString];
  NSNumber *namedValue = ZynthNamedColors[lower];
  if (namedValue) {
    NSInteger v = [namedValue integerValue];
    if (v == -1) return [UIColor clearColor];
    return [self colorFromInt:v alpha:1.0];
  }

  if ([lower hasPrefix:@"rgb"]) {
    return [self parseRGB:trimmed];
  }
  if ([lower hasPrefix:@"hsl"]) {
    return [self parseHSL:trimmed];
  }
  if ([lower hasPrefix:@"hwb"]) {
    return [self parseHWB:trimmed];
  }

  return nil;
}

+ (UIColor *)parseHex:(NSString *)hex {
  NSString *clean = [hex substringFromIndex:1];
  if (clean.length == 3) {
    NSString *r = [clean substringWithRange:NSMakeRange(0, 1)];
    NSString *g = [clean substringWithRange:NSMakeRange(1, 1)];
    NSString *b = [clean substringWithRange:NSMakeRange(2, 1)];
    clean = [NSString stringWithFormat:@"%@%@%@%@%@%@FF", r, r, g, g, b, b];
  } else if (clean.length == 4) {
    NSString *r = [clean substringWithRange:NSMakeRange(0, 1)];
    NSString *g = [clean substringWithRange:NSMakeRange(1, 1)];
    NSString *b = [clean substringWithRange:NSMakeRange(2, 1)];
    NSString *a = [clean substringWithRange:NSMakeRange(3, 1)];
    clean = [NSString stringWithFormat:@"%@%@%@%@%@%@%@%@", r, r, g, g, b, b, a, a];
  } else if (clean.length == 6) {
    clean = [clean stringByAppendingString:@"FF"];
  }

  if (clean.length != 8) return nil;
  unsigned int rgba = 0;
  [[NSScanner scannerWithString:clean] scanHexInt:&rgba];

  CGFloat r = ((rgba >> 24) & 0xFF) / 255.0;
  CGFloat g = ((rgba >> 16) & 0xFF) / 255.0;
  CGFloat b = ((rgba >> 8) & 0xFF) / 255.0;
  CGFloat a = (rgba & 0xFF) / 255.0;

  return [UIColor colorWithRed:r green:g blue:b alpha:a];
}

+ (UIColor *)parseRGB:(NSString *)input {
  NSString *clean = [self stripFunction:input];
  NSArray *parts = [clean componentsSeparatedByCharactersInSet:[NSCharacterSet characterSetWithCharactersInString:@",/"]];
  if (parts.count < 3) return nil;

  CGFloat r = [self parseComponent:parts[0] max:255];
  CGFloat g = [self parseComponent:parts[1] max:255];
  CGFloat b = [self parseComponent:parts[2] max:255];
  CGFloat a = 1.0;
  if (parts.count >= 4) {
    a = [self parseAlpha:parts[3]];
  }
  return [UIColor colorWithRed:r green:g blue:b alpha:a];
}

+ (UIColor *)parseHSL:(NSString *)input {
  NSString *clean = [self stripFunction:input];
  NSArray *parts = [clean componentsSeparatedByCharactersInSet:[NSCharacterSet characterSetWithCharactersInString:@",/"]];
  if (parts.count < 3) return nil;

  CGFloat h = [self parseHue:parts[0]];
  CGFloat s = [self parsePercent:parts[1]];
  CGFloat l = [self parsePercent:parts[2]];
  CGFloat a = 1.0;
  if (parts.count >= 4) a = [self parseAlpha:parts[3]];

  CGFloat c = (1 - fabs(2 * l - 1)) * s;
  CGFloat x = c * (1 - fabs(fmod(h / 60.0, 2) - 1));
  CGFloat m = l - c / 2;

  CGFloat r = 0, g = 0, b = 0;
  if (h < 60) {
    r = c; g = x;
  } else if (h < 120) {
    r = x; g = c;
  } else if (h < 180) {
    g = c; b = x;
  } else if (h < 240) {
    g = x; b = c;
  } else if (h < 300) {
    r = x; b = c;
  } else {
    r = c; b = x;
  }

  return [UIColor colorWithRed:r + m green:g + m blue:b + m alpha:a];
}

+ (UIColor *)parseHWB:(NSString *)input {
  NSString *clean = [self stripFunction:input];
  NSArray *parts = [clean componentsSeparatedByCharactersInSet:[NSCharacterSet characterSetWithCharactersInString:@",/"]];
  if (parts.count < 3) return nil;

  CGFloat h = [self parseHue:parts[0]];
  CGFloat w = [self parsePercent:parts[1]];
  CGFloat b = [self parsePercent:parts[2]];
  CGFloat a = 1.0;
  if (parts.count >= 4) a = [self parseAlpha:parts[3]];

  CGFloat ratio = w + b;
  if (ratio > 1) {
    w /= ratio;
    b /= ratio;
  }

  CGFloat v = 1 - b;
  CGFloat c = v - w;
  CGFloat x = c * (1 - fabs(fmod(h / 60.0, 2) - 1));

  CGFloat r = 0, g = 0, bl = 0;
  if (h < 60) {
    r = c; g = x;
  } else if (h < 120) {
    r = x; g = c;
  } else if (h < 180) {
    g = c; bl = x;
  } else if (h < 240) {
    g = x; bl = c;
  } else if (h < 300) {
    r = x; bl = c;
  } else {
    r = c; bl = x;
  }

  return [UIColor colorWithRed:r + w green:g + w blue:bl + w alpha:a];
}

+ (NSString *)stripFunction:(NSString *)input {
  NSRange open = [input rangeOfString:@"("];
  NSRange close = [input rangeOfString:@")" options:NSBackwardsSearch];
  if (open.location == NSNotFound || close.location == NSNotFound || close.location <= open.location) {
    return input;
  }
  return [[input substringWithRange:NSMakeRange(open.location + 1, close.location - open.location - 1)]
      stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
}

+ (CGFloat)parseHue:(NSString *)value {
  NSString *trimmed = [value stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
  if ([trimmed hasSuffix:@"deg"]) {
    trimmed = [trimmed substringToIndex:trimmed.length - 3];
  }
  CGFloat h = [trimmed doubleValue];
  if (h < 0) h = fmod(h, 360) + 360;
  return fmod(h, 360);
}

+ (CGFloat)parsePercent:(NSString *)value {
  NSString *trimmed = [value stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
  if ([trimmed hasSuffix:@"%"]) {
    trimmed = [trimmed substringToIndex:trimmed.length - 1];
  }
  CGFloat v = [trimmed doubleValue] / 100.0;
  return fmax(0.0, fmin(1.0, v));
}

+ (CGFloat)parseComponent:(NSString *)value max:(CGFloat)max {
  NSString *trimmed = [value stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
  if ([trimmed hasSuffix:@"%"]) {
    trimmed = [trimmed substringToIndex:trimmed.length - 1];
    return fmax(0.0, fmin(1.0, [trimmed doubleValue] / 100.0));
  }
  return fmax(0.0, fmin(1.0, [trimmed doubleValue] / max));
}

+ (CGFloat)parseAlpha:(NSString *)value {
  NSString *trimmed = [value stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
  if ([trimmed hasSuffix:@"%"]) {
    trimmed = [trimmed substringToIndex:trimmed.length - 1];
    return fmax(0.0, fmin(1.0, [trimmed doubleValue] / 100.0));
  }
  return fmax(0.0, fmin(1.0, [trimmed doubleValue]));
}

+ (UIColor *)colorFromInt:(NSInteger)rgb alpha:(CGFloat)alpha {
  CGFloat r = ((rgb >> 16) & 0xFF) / 255.0;
  CGFloat g = ((rgb >> 8) & 0xFF) / 255.0;
  CGFloat b = (rgb & 0xFF) / 255.0;
  return [UIColor colorWithRed:r green:g blue:b alpha:alpha];
}

@end
